### PR TITLE
Add a runTests function

### PR DESCRIPTION
**Pull request by erikd**
And actually use it in the Hedhehog tests.

Closes: https://github.com/hedgehogqa/haskell-hedgehog/issues/97

@thumphries @jystic @nhibberd

### CODE REVIEW

File: hedgehog/test/test.hs [hunk 1]
```diff
@@ -1,18 +1,10 @@
...
 main :: IO ()
-main = do
-  hSetBuffering stdout LineBuffering
-  hSetBuffering stderr LineBuffering
-
-  results <- sequence [
-      Test.Hedgehog.Text.tests
+main =
+  runTests
+    [ Test.Hedgehog.Text.tests
```
Comment (nhibberd) [inline hedgehog/test/test.hs]: Probably good to keep the same syntax style here, start of list on previous line that is

Comment (erikd) [reply hedgehog/test/test.hs]: :+1: Fixed the formatting in the comment for `runTests` as well.

**Comment by thumphries** [comment]
I feel a bit weird exporting a function that calls `exitWith` from the top level, especially with an innocuous name like `runTests`. Wondering if it would make more sense from `Hedgehog.Main` or something like that? Or a name with `suite` or `main` in it?

(not requesting changes per se, just wondering how you two feel about it)

**Comment by erikd** [comment]
I'm not married to the name, I just don't want to have to redefine this function in every test suite I write. I also think what you said makes a lot of sense. So how about:

* We rename it `runTestsuie`.
* Move it to `Hedgehog.Main` and not export if from `Hegdehog` requiring people to explicitly import `Hedgehog.Main`.
* Since if calls `exitFailure` on any failure, maybe it should also call `exitSuccess` on success. Ie this function *always* terminates.

**Comment by erikd** [comment]
Hmm, no that I think about it, I'm mot so sure about the last part. I could imagine wanting to do:
```
   runTests
        [ Test.Wibble.One
        ]
  -- Only run the following tests if the previous one passed.
   runTests
        [ Test.Wibble.Two
        , Test.Wibble.Three
        ]
```

**Comment by moodmosaic** [comment]
>Move it to Hedgehog.Main and not export if from Hegdehog requiring people to explicitly import Hedgehog.Main.

Hmm... Isn't this a little bit weird? Perhaps it could be exposed as 'internal' from Hedgehog itself so we keep importing only the essential stuff:

```haskell
import           Hedgehog
import qualified Hedgehog.Gen as Gen
import qualified Hedgehog.Range as Range
```

**Comment by erikd** [comment]
The idea of putting this in a separate `Hedgehog.Main` is growing on me. Maybe we should also provide more than one runner. `Hedgehog.Main`might provide fhe functions:

```
setLineBufferring :: IO ()

-- Run the tests and return `True` if they all passed, `False` otherwise.
runTestsuite :: [IO Bool] -> IO Bool

-- If any test fails, exit with `ExitFailure`.
runTestsuiteExitOnFail :: [IO Bool] -> IO ()

-- This _always_ exits, `ExitFailure` if any test fails, `ExitSuccess` otherwise.
runTestsuiteExit :: [IO Bool] -> IO a

```

**Comment by moodmosaic** [comment]
Are we 100% sure about the `Main` part? Not that I can currently come up with an alternative...

**Comment by erikd** [comment]
@moodmosaic The import you have in your example are the ones you would put in an actual test suite.. I usually keep actuall tests out of the top level test runner. With this new functionality, my top level test runner would look like:

```
import qualified Hedgehog.Main as HMain

import qualified Test.Wibble.One
import qualified Test.Wibble.Two
import qualified Test.Wibble.Three

main :: IO ()
main = do
  HMain.setLineBuffering
  HMain.runTestsuiteExit
    [ Test.Wibble.One
    , Test.Wibble.Two
    , Test.Wibble.Three
    ]
   
I'm not married to the name `Main`. Might be `Run` or `Runner` instead. Or maybe with these new names and type signatures we can just export them from `Hedgehog`.

**Comment by gwils** [comment]
I don't like the name Hedgehog.Main, but Run or Runner I could get behind. It just feels a bit wrong to me for a library (not an executable) to export a module called Main. I would prefer that this be exported from Hedgehog rather than a separate module regardless.

My preference is the version that calls `exitFailure` on failure but _does not_ call `exitSuccess` on success. As Erik has pointed out, you may wish to call many test runs sequentially.

**Comment by erikd** [comment]
> My preference is the version that calls exitFailure on failure but does not call exitSuccess on success.

This is why my (strawman) proposal now has three versions of the runner function :).

File: hedgehog/src/Hedgehog/Internal/Runner.hs [hunk 2]
```diff
@@ -394,6 +401,35 @@ checkParallel =
           Nothing
       }
 
+-- | Run list of tests.
+--
+-- Given a number of separate modules all defining test funtion named `tests`,
+-- the individual module level tests can be collected together and run in
+--  sequence (one module of tests at a time) and the funcion will will exit
```
Comment (moodmosaic) [inline hedgehog/src/Hedgehog/Internal/Runner.hs]: Nitpick: there seems to be an extra space between `sequence`.
Nitpick: `will` occurs two times in a row on that sentence.

Comment (erikd) [reply hedgehog/src/Hedgehog/Internal/Runner.hs]: Yeah, this PR will not go ahead as it is anyway.

**Comment by thumphries** [comment]
Closing this based on Slack discussions we had back in March.

The decision was to not rush into exporting a top-level runner, since public exports are forever and we all had some reservations about the current design. There is also the question of terminal encoding, Windows, etc to consider.

File: hedgehog/src/Hedgehog/Internal/Runner.hs [hunk 3]
```diff
@@ -394,6 +401,22 @@ checkParallel =
...
+  fmap and . sequence
+
+-- | Like `runTestsuite` but exit with `exitFailure` if one or more of the
+-- tests fail.
+
+runTestsExitFail :: [IO Bool] -> IO ()
+runTestsExitFail tests = do
+  result <- runTests tests
+  unless result
+    exitFailure
```
Comment (jystic) [inline hedgehog/src/Hedgehog/Internal/Runner.hs]: If you rename this to `runTests` and delete the other function I'll be happy

File: hedgehog/test/test.hs [hunk 4]
```diff
@@ -1,18 +1,10 @@
...
-import           System.Exit (exitFailure)
+import           Hedgehog (runTestsExitFail)
 
 import qualified Test.Hedgehog.Text
 
 
 main :: IO ()
-main = do
-  hSetBuffering stdout LineBuffering
-  hSetBuffering stderr LineBuffering
```
Comment (jystic) [inline hedgehog/test/test.hs]: Don't we need this? Should probably be in `runTests`

https://github.com/ambiata/disorder.hs/blob/master/disorder-core/src/Disorder/Core/Main.hs#L26-L28

**Comment by jystic** [comment]
💯

**Pull request merged**